src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -928,19 +900,316 @@ int xocl_cleanup_mem(struct xocl_drm *drm_p)
 	return ret;
 }
 
+static int xocl_cleanup_drm_memory_manager(struct xocl_mm *xocl_mm)
+{
+	int i = 0;
+
+	if (!xocl_mm)
+		return 0;
+
+	for (i = 0; i < xocl_mm->m_count; i++) {
+		if (xocl_mm->mm_usage_stat && xocl_mm->mm_usage_stat[i])
+			vfree(xocl_mm->mm_usage_stat[i]);
+	}
+
+	vfree(xocl_mm->mm_usage_stat);
+
+	if (xocl_mm->bo_usage_stat)
+		vfree(xocl_mm->bo_usage_stat);
+
+	if (xocl_mm->mm)
+		vfree(xocl_mm->mm);
+
+	vfree(xocl_mm);
+
+	return 0;
+}
+
+static int xocl_init_drm_mm(struct xocl_drm *drm_p, uint64_t mm_start_addr,
+		uint64_t mm_end_addr, uint32_t m_count)
+{
+	struct xocl_mm *xocl_mm = drm_p->xocl_mm;
+	size_t size = 0;
+	int err = 0;
+	int i = 0;
+	
+	BUG_ON(!mutex_is_locked(&drm_p->mm_lock));
+
+	if (!xocl_mm)
+		return -EINVAL;
+
+	size = m_count * sizeof(void *);
+	xocl_mm->mm_usage_stat = vzalloc(size);
+	if (!xocl_mm->mm_usage_stat) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	/* Initialize memory status for the memory manager */
+	xocl_mm->bo_usage_stat = vzalloc(XOCL_BO_USAGE_TOTAL *
+			sizeof(struct drm_xocl_mm_stat));
+	if (!xocl_mm->bo_usage_stat) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	for (i = 0; i < m_count; i++) {
+		xocl_mm->mm_usage_stat[i] = vzalloc(sizeof(struct drm_xocl_mm_stat));
+		if (!xocl_mm->mm_usage_stat[i]) {
+			err = -ENOMEM;
+			goto error;
+		}
+	}
+
+	/* Initialize the memory manager */
+	xocl_mm->mm = vzalloc(sizeof(struct drm_mm));
+	if (!xocl_mm->mm) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	drm_mm_init(xocl_mm->mm, mm_start_addr, (mm_end_addr - mm_start_addr));
+	if (!xocl_mm->mm) {
+		xocl_err(drm_p->ddev->dev,
+				"init memory manager failed");
+		err = -EINVAL;
+		goto error;
+	}
+
```

Comment:
This is my last comment in the older PR.
"Line 966 already check if xocl_mm->mm is NULL. Why check this again? Do you mean drm_mm_init() will free and set xocl_mm->mm to NULL if initialized incorrectly?"

Reply:
We may not need to check again. drm_mem_init doesn't set anything and not return any error code. 
We should continue without any check.